### DISCUSSION TITLE
How to use set_page method to change the page ?

### DISCUSSION THREAD

**Discussion by am077** [writer/writer-framework#691] [Q&A] [answered]
I'm flowing this page https://dev.writer.com/framework/page-routes but that doesn't seems work. I have used it in a handler. The UI is completely CMC. Is anything I'm missing ?

```
def upload_page(state, ui):
    state.set_page("upload")

def settings_page(state, ui):
    state.set_page("settings")
```

**Comment by mmikita95** [comment, marked as answer]
Hello @am007! Mind sharing a snippet of your UI code? One hint I would suggest is to check whether the CMC pages have a defined `key` in `content` dictionary in addition to `id` component property – it is slightly counterintuitive, but pages would not be accessible otherwise, as `id` is what we use to "store" it on backend, and `key` is what frontend uses to manage display.

As an example, this won't work:
```py
with wf.init_ui() as ui:
    for page in range(1, 4):
        with ui.Page(id=f"test-cmc-{page}"):
            ui.Text({"text": f"Hello Page {page}!"})
            for page_link in range(1, 4):
                if page_link == page:
                    continue
                ui.Button({"text": f"Go to Page {page_link}"}, handlers={
                    "wf-click": {1: page_1, 2: page_2, 3: page_3}[page_link]
                })
```
but this will:
```py
with wf.init_ui() as ui:
    for page in range(1, 4):
        with ui.Page(content={"key": f"test-cmc-{page}"}, id=f"test-cmc-{page}"):
            ui.Text({"text": f"Hello Page {page}!"})
            for page_link in range(1, 4):
                if page_link == page:
                    continue
                ui.Button({"text": f"Go to Page {page_link}"}, handlers={
                    "wf-click": {1: page_1, 2: page_2, 3: page_3}[page_link]
                })
```

**Reply by am077** [reply]
I didn't know about **content** dictionary. That worked. Thanks a lot.

**Reply by mmikita95** [reply]
No problem! Glad I could help. Don't hesitate to reach out if there's anything else you need help with!
